### ISSUE TITLE
Add a way to convert from an `Interface` to `Option<Interface>`

### ISSUE THREAD

**Issue by jrmuizel** [microsoft/windows-rs#1339]
Some APIs like [`VSSetConstantBuffers`](https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Graphics/Direct3D11/struct.ID3D11DeviceContext.html#method.VSSetConstantBuffers) take an `*const Option<ID3D11Buffer>`. Given a `&ID3D11Buffer` I'd like to be able to construct a `&Option<ID3D11Buffer>` to pass to a function like this.

**Comment by PhoNoTheSky** [comment]
Doing `&buffer.clone().into()` works fine for turning `&ID3D11Buffer` into `&Option<ID3D11Buffer>`

**Comment by kennykerr** [comment]
This API is difficult to use because it really expects an array but the Rust bindings aren't yet adapted to accept a slice. 

This is a duplicate of #479 that I hope to get to soon.

**Comment by kennykerr** [comment]
Sorry for the delay, #1562 adds support for Win32 array parameters.

**Comment by chyyran** [comment]
Unfortunately #1562 does not solve the core issue. Even for functions that take slices, they expect slices of `&[Option<T>]`, which makes it difficult when you have a `&[T]`. The `.clone().into()` trick works well if you have a slice of size 1, but for slices without known size, we either have to map over the slices of `T` to `Option<T>::Some`, or transmute `T` to `Option<T>`/

**Comment by kennykerr** [comment]
Reopening to address this.

**Comment by kennykerr** [comment]
This function is generated as follows:

```rust
pub unsafe fn VSSetConstantBuffers(&self, startslot: u32, ppconstantbuffers: &[::core::option::Option<ID3D11Buffer>]) {
    (::windows::core::Interface::vtable(self).VSSetConstantBuffers)(::core::mem::transmute_copy(self), ::core::mem::transmute(startslot), ppconstantbuffers.len() as _, ::core::mem::transmute(::windows::core::as_ptr_or_null(ppconstantbuffers)))
}
```

This is in line with the metadata:

```C#
unsafe void VSSetConstantBuffers([In] uint StartSlot, [In] uint NumBuffers, [Optional][In][NativeArrayInfo(CountParamIndex = 1)] ID3D11Buffer* ppConstantBuffers);
```

Unfortunately, there's nothing in the metadata that says that the interface pointers in the array may not be null, hence the `Option`.

**Comment by kennykerr** [comment]
Ah that's a good point, thanks for the reminder! Yes, I'll experiment with using `IntoParam` for input arrays.

**Comment by aconverse** [comment]
I know this issue is very old. I can open a new issue if that is preferred. 

But this was closed with a change (https://github.com/microsoft/windows-rs/pull/2233) that had to be reverted and never re-opened. 

Going back to the original use case: User has `&ID3D11Buffer` and wants to call `VSSetConstantBuffers()`. 

Can we get an AsRef to take us from `&ID3D11Buffer` to a `&Option<ID3D11Buffer>`? (Any more generically 
this transformation for any COM object). 

Right now the options seem to be:
- Call `transmute` from application code
- Clone the `ID3D11Buffer` churning ref counts

See also: https://github.com/microsoft/windows-rs/issues/2930#issuecomment-2005168375

**Comment by zopsicle** [comment]
If VSSetConstantBuffers took Option\<[Ref](https://docs.rs/windows-core/latest/windows_core/struct.Ref.html)\<ID3D11Buffer>> instead, neither clone nor transmute would be necessary. That would be ideal IMO.

**Comment by kennykerr** [comment]
Yes, I added `Ref` to deal with such cases. Just need to apply it more broadly.

**Comment by kennykerr** [comment]
The issue is valid, but solving it broadly adds disproportionate complexity and hidden behavior. The existing `&[Option<T>]` projection is the most practical low-level representation that I can see. I experimented with using `Ref` and param slices but it becomes overly complex with diminishing returns. Your best bet is to wrap these low-level API calls in safe and idiomatic wrapper libraries like [windows-canvas](https://github.com/microsoft/windows-rs/blob/master/docs/crates/windows-canvas.md).
